### DISCUSSION TITLE
regarding `classNames` function in the documentation, dynamic classes, class concatenation

### DISCUSSION THREAD

**Discussion by SMKH-PRO** [tailwindlabs/tailwindcss#9539] [Help] [answered]
Dear Tailwind Community,

We're using tailwind css in our projects and we use `classNames` functions as per documentation,

But I am concerned about the performance, since this function using filter loop and join operation and is being called inside the JSX (which means it will be called on each render, right?), loop on each render? won't it increase render time of the component? and eventually will cause performance issues?

we used this in big projects with a-lot of components and now I am concern that it is going to effect render time of components.

We're currently not facing such performance issues but very concerned about this because our project is constantly growing and this helper function is being used on every component for classNames, I am concerned it must be increasing render time and causing un-noticeable performance issues which will later become noticeable as the project grows.

I would really love community's input on this.

<img width="1620" alt="Screenshot 2022-10-13 at 3 52 32 PM" src="https://user-images.githubusercontent.com/41765372/195578887-fcd376bf-f51d-4b89-becf-32e01335f6fc.png">

<img width="1602" alt="Screenshot 2022-10-13 at 4 01 17 PM" src="https://user-images.githubusercontent.com/41765372/195580132-a6a47fc0-9276-4e00-94aa-64e2cb821e9f.png">

Thank you!

**Comment by RobinMalfait** [comment, marked as answer]
Hey! Thank you for your question!
Much appreciated! 🙏

Whenever we talk about "optimizations" you have to keep in mind that everybody optimizes for different reasons (and it's also context dependant). Sometimes performance is the most critical part for a specific piece of code of the application. If that's the case, often (not always) readability and maintainability gets sacrificed just to squeeze out those extra nanoseconds. Sometimes people sacrifice correctness over performance. For example: look at this `Math.sqrt(n)` implementation they used in Quake, called the `Fast inverse square root` created by John Carmack. It's an estimation that is _good enough_ but it is _fast_.

```cp
float Q_rsqrt( float number )
{
	long i;
	float x2, y;
	const float threehalfs = 1.5F;

	x2 = number * 0.5F;
	y  = number;
	i  = * ( long * ) &y;                       // evil floating point bit level hacking
	i  = 0x5f3759df - ( i >> 1 );               // what the fuck? 
	y  = * ( float * ) &i;
	y  = y * ( threehalfs - ( x2 * y * y ) );   // 1st iteration
//	y  = y * ( threehalfs - ( x2 * y * y ) );   // 2nd iteration, this can be removed

	return y;
}
```
– https://en.wikipedia.org/wiki/Fast_inverse_square_root

---

Whenever you are concerned of anything that might have a (negative) performance impact then the most important thing you have to do is: `measure`.

Whenever you measure (let's say at least ~30 times times to get a more statistically correct result) you also have to make sure that you don't focus on this one particular function on its own. You have to look at it in a more broad way and see where your _bottleneck_ is. 
It doesn't make sense to optimize the `classNames` function to go from a total of 3ms to 1ms if there are functions that take over 100ms in the same render cycle. That's a waste of time to even try and optimize it.

You are right that the `classNames`:
- Is another function call.
- Uses a `filter` and a `join` which technically loops over the array twice (maybe more or less depending on internal implementation).
- Is used multiple times in multiple places over multiple renders.

It's true that therefore the `classNames` function has, in the worst case scenario (nothing is filtered out and therefore you have to loop twice because of the 1. filter and the 2. join), a runtime cost of `O(2n)`. Now this looks really bad, however it is rarely that your `n` exceeds a handful of elements in the array. In your screenshot above there are only 2 elements for example.


Here is a bit of benchmarking code for you:
```js
function classNames(...classes) {
  return classes.filter(Boolean).join(' ')
}

// With classNames
for (let i = 0; i < 30; i++) {
  console.time('With classNames')
  for (let i = 0; i < 1_000_000; i++) {
    let className = classNames(
      i === 0 && 'border-t',
      i % 100 === 0 && 'border-blue-100',
      'border-b border-gray-200 p-4'
    )
  }
  console.timeEnd('With classNames')
}
```

The TL;DR of this snippet is that we are timing (30 times in a row) how long it takes for the `className` to be calculated 1 million times (`1 000 000`).

These are the results:
```
With classNames: 50.547ms
With classNames: 48.766ms
With classNames: 45.467ms
With classNames: 46.321ms
With classNames: 45.414ms
With classNames: 45.805ms
With classNames: 45.346ms
With classNames: 45.615ms
With classNames: 45.395ms
With classNames: 46.071ms
With classNames: 45.845ms
With classNames: 45.401ms
With classNames: 45.564ms
With classNames: 45.408ms
With classNames: 45.955ms
With classNames: 45.927ms
With classNames: 45.425ms
With classNames: 45.511ms
With classNames: 46.342ms
With classNames: 46.009ms
With classNames: 45.374ms
With classNames: 45.48ms
With classNames: 46.119ms
With classNames: 45.753ms
With classNames: 45.496ms
With classNames: 47.22ms
With classNames: 45.495ms
With classNames: 46.248ms
With classNames: 45.649ms
With classNames: 46.037ms
```

~45ms on average seem like a lot, but you have to keep in mind that you are doing this 1 million times  for each call. We did this for 30 times so a total of 30 million calls (`30 000 000`). The total of those 30 million calls is about 1.380s which means that each call on its own takes about `46 nanoseconds`. Also note that this is in **nano**seconds, not **milli**seconds that most web developers consider Blazingly Fast™. If, after measuring, this is your biggest bottleneck of your whole application then:

1. Good news! You can optimize it by constructing the `className` yourself manually.
3. Congratulations, you have the most performant web application on planet earth 🥳

Now, for the manual approach, you can use something like this:
```js
// Without classNames
for (let i = 0; i < 30; i++) {
  console.time('Without classNames')
  for (let i = 0; i < 1_000_000; i++) {
    let className = `${i === 0 ? 'border-t' : ''}${
      i % 100 === 0 ? 'border-blue-100' : ''
    } border-b border-gray-200 p-4`
  }
  console.timeEnd('Without classNames')
}
```
... it's not as nicely formatted and gets a bit overwhelming the moment you have more conditions. Prettier formatted it like this for me. (Hence why we added the `classNames` function to improve readability and maintainability).

The results of running this are interesting:
```
Without classNames: 2.313ms
Without classNames: 2.595ms
Without classNames: 1.205ms
Without classNames: 1.206ms
Without classNames: 1.2ms
Without classNames: 1.286ms
Without classNames: 1.278ms
Without classNames: 1.241ms
Without classNames: 1.208ms
Without classNames: 1.205ms
Without classNames: 1.204ms
Without classNames: 1.222ms
Without classNames: 1.204ms
Without classNames: 1.204ms
Without classNames: 1.24ms
Without classNames: 1.207ms
Without classNames: 1.213ms
Without classNames: 1.229ms
Without classNames: 1.251ms
Without classNames: 1.264ms
Without classNames: 1.213ms
Without classNames: 1.209ms
Without classNames: 1.204ms
Without classNames: 1.204ms
Without classNames: 1.209ms
Without classNames: 1.204ms
Without classNames: 1.205ms
Without classNames: 1.204ms
Without classNames: 1.208ms
Without classNames: 1.204ms
```

It's way way way faster, but again, this is done 1 million times for each call. These 30 million calls took a total of `39.783ms` which means that each call took `1.3261 nanoseconds`. Again, this is indeed 40x faster than the previous implementation but realistically this doesn't even matter in the grand scheme of things.

Here is the full benchmarking code in case you want to run it yourself:
```js
function classNames(...classes) {
  return classes.filter(Boolean).join(' ')
}

// With classNames
console.time('With classNames')
for (let i = 0; i < 30; i++) {
  console.time(' - run')
  for (let i = 0; i < 1_000_000; i++) {
    let className = classNames(
      i === 0 && 'border-t',
      i % 100 === 0 && 'border-blue-100',
      'border-b border-gray-200 p-4'
    )
  }
  console.timeEnd(' - run')
}
console.timeEnd('With classNames')

// Without classNames
console.time('Without classNames')
for (let i = 0; i < 30; i++) {
  console.time(' - run')
  for (let i = 0; i < 1_000_000; i++) {
    let className = `${i === 0 ? 'border-t' : ''}${
      i % 100 === 0 ? 'border-blue-100' : ''
    } border-b border-gray-200 p-4`
  }
  console.timeEnd(' - run')
}
console.timeEnd('Without classNames')
```

---

So to summarise:

1. Measure
2. Measure
3. Measure
4. Yes, it is technically a bit slower, but I can almost guarantee that this is not an issue in the real world at all.
5. Measure

**Comment by SMKH-PRO** [comment]
@RobinMalfait 
Thank you so much for an explained answer, I am very satisfied with this now.  😊

**Comment by kimskogsmo** [comment]
If for some people this excellent explanation by @RobinMalfait  still doesn't satisfy their need for performance, there are faster alternatives to the "classnames" package, like "clsx" and "classix", which are up to 5x as fast, although with somewhat less flexibility as a drawback.
